### PR TITLE
image: check kernel/gadget publisher vs model brand, warn on store disconnected snaps 

### PR DESCRIPTION
**Pull request by pedronis**

### CODE REVIEW

File: image/helpers.go [hunk 1]
```diff
@@ -108,16 +109,28 @@ func StoreAssertionFetcher(sto Store, dlOpts *DownloadOptions, db *asserts.Datab
...
 	// cross checks
-	return snapasserts.CrossCheck(info.Name(), sha3_384, size, &info.SideInfo, db)
+	if err := snapasserts.CrossCheck(info.Name(), sha3_384, size, &info.SideInfo, db); err != nil {
+		return nil, err
+	}
+
+	a, err := db.Find(asserts.SnapDeclarationType, map[string]string{
+		"series":  release.Series,
+		"snap-id": info.SnapID,
+	})
```
Comment (zyga) [inline image/helpers.go]: I'm not sure how various checks here operate, please bear with me...
 - we hash the actual .snap file
 - we fetch and check (presumably that the signature itself is correct) all the assertions affecting the snap with the hash we've got
 - we cross-check that the snap name and snap hash are correct and match side info

...and now we come to the new code:
 - we look up a snap declaration that matches our current series and the given snap-id

But isn't this exactly what is done in CrossCheck (apart from the fact that we also have the assertion and can return it)?

File: image/image.go [hunk 2]
```diff
@@ -312,18 +329,33 @@ func bootstrapToRootDir(sto Store, model *asserts.Model, opts *Options, local *l
...
 				return err
 			}
+			var kind string
+			switch typ {
+			case snap.TypeKernel:
+				kind = "kernel"
+			case snap.TypeGadget:
+				kind = "gadget"
+			}
+			if kind != "" { // kernel or gadget
```
Comment (zyga) [inline image/image.go]: Is this the essential part of the change? That the publisher of the kernel and gadget snap matches the declaration in the model assertion (or that it is a snap from canonical)?

Comment (pedronis) [inline image/image.go]: yes, this is the essential part of the change

File: image/helpers.go (hunk 1 above)
Comment (pedronis) [inline image/helpers.go]: yes, we just get hold of the cross-verified snap-declaration and return it (CrossCheck is used also in snapd itself and changing its signature doesn't make as much sense), anyway the style is common also inside snapd,  a main task verify the assertions, later tasks get hold of the snap-declaration (from the db) later to do their work (this is just done very close together)

File: image/helpers.go [hunk 3]
```diff
@@ -108,16 +109,28 @@ func StoreAssertionFetcher(sto Store, dlOpts *DownloadOptions, db *asserts.Datab
...
+		return nil, fmt.Errorf("cannot fetch snap signatures/assertions: %v", err)
 	}
 
 	// cross checks
-	return snapasserts.CrossCheck(info.Name(), sha3_384, size, &info.SideInfo, db)
+	if err := snapasserts.CrossCheck(info.Name(), sha3_384, size, &info.SideInfo, db); err != nil {
+		return nil, err
+	}
+
+	a, err := db.Find(asserts.SnapDeclarationType, map[string]string{
```
Comment (mvo5) [inline image/helpers.go]: Nice!

File: image/image.go [hunk 4]
```diff
@@ -120,6 +122,10 @@ func Prepare(opts *Options) error {
 		return err
 	}
 
+	// FIXME: limitation until we can pass series parametrized much more
+	if model.Series() != release.Series {
```
Comment (mvo5) [inline image/image.go]: :+1:

File: image/image.go [hunk 5]
```diff
@@ -312,18 +329,33 @@ func bootstrapToRootDir(sto Store, model *asserts.Model, opts *Options, local *l
...
+			var kind string
+			switch typ {
+			case snap.TypeKernel:
+				kind = "kernel"
+			case snap.TypeGadget:
+				kind = "gadget"
+			}
+			if kind != "" { // kernel or gadget
+				publisher := snapDecl.PublisherID()
+				if publisher != model.BrandID() && publisher != "canonical" {
```
Comment (mvo5) [inline image/image.go]: We have similar logic in devicemgr.go - I guess its all pretty trivial currently so not a big deal. But it does mean we need to change two places if we ever want to change this policy. I wonder if its worth having a shared helper?

File: image/image.go [hunk 6]
```diff
@@ -346,6 +378,9 @@ func bootstrapToRootDir(sto Store, model *asserts.Model, opts *Options, local *l
 			Unasserted: info.SnapID == "",
 		})
 	}
+	if len(locals) > 0 {
```
Comment (mvo5) [inline image/image.go]: Nice!

File: image/image.go (hunk 5 above)
Comment (pedronis) [inline image/image.go]: if the logic gets more complicated we could (I can add a question TODO?) , atm is a 3 lines check (and we would need a place to put it)

**Pull request merged**